Adding SimEnergyDeposit filter module for systematics studies (mostly geared towards ICARUS but also general to both detectors)

jzettle opened the pull request:
`A request that this PR is really geared towards a patch release of sbncode v09_89_01_01 for ICARUS systematics. We can include it here for continued functionality in develop, but I need a patch release (github doesn't let me do it or I am not git-savvy enough to have found it quickly).`

### Description 
Please provide a detailed description of the changes this pull request introduces. If available, also link to a docdb link where the issue/change have been presented on/discussed.
This pull request introduces a larsoft module that can be run during the detsim stage to filter out SimEnergyDeposits around a rectangular volume and creates a new SimEnergyDeposit collection (with a different name, `sedfilter`) that can be passed to the downstream WireCell processing. The goal of this is to allow for a systematic variation sample that addresses the fact that the induction 1 wire gap at z = 0 is not simulated in the icarus geometry. **This PR does not change anything with the standard ICARUS or SBND simulation and the module is completely optional to include**. 

This PR addresses the module and a companion PR in icaruscode (https://github.com/SBNSoftware/icaruscode/pull/778) addresses the fcl configuration and WireCell configuration options of running it as a variation sample. 

- [X] Have you added a label? (bug/enhancement/physics etc.)
- yes
- [X] Have you assigned at least 1 reviewer?
- yes
- [X] Is this PR related to an open issue / project?
- yes, for ICARUS systematics studies for the first oscillation analysis, https://github.com/SBNSoftware/icaruscode/issues/769
- [X] Does this PR affect CAF data format? If so, please assign a CAF maintainer as additional reviewer.
- no
- [X] Does this PR require merging another PR in a different repository (such as sbnanobj/sbnobj etc.)? If so, please link it in the description.
- yes, there is a companion PR for icaruscode that includes fcl files and wirecell configuration options in order to run this as a systematic variation
- [X] Are you submitting this PR on behalf of someone else who made the code changes? If so, please mention them in the description.
- no, this is largely my own work with discussions with @jzennamo

PetrilloAtWork commented on sbncode/DetSim/fcl/icarus_simedepfilter.fcl:
Please add references to why these values are chosen, and maybe the units (cm) in some comment.
Also, the parameter `OutSimEnergyDepositLabel` is not used by the module and should be removed.

PetrilloAtWork commented on sbncode/DetSim/fcl/icarus_simedepfilter.fcl:
```diff
@@ -0,0 +1,43 @@
+  P2_Z:	900
+}
+
+filter_ind1gap: @local::simedepfilter_ind1gap
+filter_cathodewest: @local::simedepfilter_cathodewest
+filter_cathodeeast: @local::simedepfilter_cathodeeast
```
What is the idea behind these aliases?

jzettle replied:
Mostly ignorance, don't need these can use the original names in the same file, removed

SFBayLaser reviewed (commented):
Gianluca has made a number of detailed comments which all look like either very good suggestions or important questions. So I will be good when these are resolved.

jzettle commented:
Thanks @PetrilloAtWork for the detailed comments and suggestions! I will work on those and ping again once finished. It should transpose the commits onto PR #488 which is the production release version of this PR.

jzennamo commented:
@PetrilloAtWork >> @jzennamo, aside from the one comment I had I didn't have anything to add beyond what @PetrilloAtWork already added.

jzettle commented:
Okay I have addressed the comments but it says that develop got merged into my branch, which I did not want to do locally to preserve my own version. I cannot push the current version of my branch due to conflicts now associated with that merge so I will need to find a workaround

jzettle commented:
Okay I have addressed @PetrilloAtWork's comments here but now I am concerned about putting this in v09_89_01_01 specifically now that this branch is merged with develop but someone can alleviate my concerns if they are unfounded

PetrilloAtWork commented on sbncode/DetSim/FilterSimEnergyDeposits_module.cc:
```diff
@@ -50,55 +41,38 @@ class FilterSimEnergyDeposits : public art::EDProducer {
+  , fBox{p.get<double>("P1_X"), p.get<double>("P2_X"),
+	 p.get<double>("P1_Y"), p.get<double>("P2_Y"),
+	 p.get<double>("P1_Z"), p.get<double>("P2_Z")}
+  , fA{p.get<double>("A", 0.0)}
+  , fB{p.get<double>("B", 0.0)}
+  , fC{p.get<double>("C", 0.0)}
```
Please remove the default values: with these values the module will not work correctly (divide by 0), and it may give the illusion to the users that no configuration is needed.

PetrilloAtWork commented on sbncode/DetSim/FilterSimEnergyDeposits_module.cc:
```diff
@@ -50,55 +41,38 @@ class FilterSimEnergyDeposits : public art::EDProducer {
-  double ShiftX(double z);
+  geo::BoxBoundedGeo fBox;
+  static constexpr auto kModuleName = "FilterSimEnergyDeposit";
+  const double ShiftX(double z);
```
This should be:
```suggestion
  double ShiftX(double z) const;
```
The former promises to return a "constant" value, while the latter promises that it's not going to change anything in the whole object (that is, treat it like `const FilterSimEnergyDeposits`).

PetrilloAtWork commented on sbncode/DetSim/FilterSimEnergyDeposits_module.cc:
```diff
@@ -50,55 +41,38 @@ class FilterSimEnergyDeposits : public art::EDProducer {
-}
 
-double FilterSimEnergyDeposits::ShiftX(double z)
+const double FilterSimEnergyDeposits::ShiftX(double z)
```
As above:
```suggestion
double FilterSimEnergyDeposits::ShiftX(double z) const
```

PetrilloAtWork commented on sbncode/DetSim/FilterSimEnergyDeposits_module.cc:
```diff
@@ -50,55 +41,38 @@ class FilterSimEnergyDeposits : public art::EDProducer {
-  double a = -0.431172;
-  double b = -2.52724;
-  double c = 0.22043;
+  double a = fA;
+  double b = fB;
+  double c = fC;
```
You can also use the three values directly below. Your choice.

PetrilloAtWork commented on sbncode/DetSim/fcl/icarus_simedepfilter.fcl:
```diff
@@ -3,41 +3,16 @@ BEGIN_PROLOG
+  P2_X: 400 #cm, create a box that spans the full x,y range. Not super-precise.
+  P2_Y: 200 #cm, create a box that spans the full x,y range. Not super-precise.
+  P2_Z: 2 #cm, other edge of the filtered gap based on the 37 mm total size of the wire gap
+  A: -0.431172
+  B: -2.52724
+  C: 0.22043
```
Add the comment to where these come from:
```suggestion
  # from SBN DocDB 38701:
  A: -0.431172
  B: -2.52724
  C: 0.22043
```

PetrilloAtWork reviewed (changes requested):
Almost there.
I request a couple of changes more:
 * add in the comment where the parameter values are from
 * fix the "constantness" of `ShiftX()` method

PetrilloAtWork reviewed (approved):
Good for me, and thank you for putting up with my comments.
